src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2489,6 +2489,13 @@ internal override void GetDiagnostics(CompilationStage stage, bool includeEarlie
                             {
                                 throw ExceptionUtilities.Unreachable;
                             }
+                            catch (OperationCanceledException e) when (cancellationToken.IsCancellationRequested && e.CancellationToken != cancellationToken)
```

Comment:
Rather than repeating this code at every `Parallel.For` in the code base did you consider writing a new helper, `RoslynParallel.For`, centralize the work around there and have everyone use that helper instead?

Reply:
Implemented this approach in the second commit.